Compatibility issue with go 1.16

Signing message on go version 1.16 makes crypto/hmac  lib panic

```
package main

import (
        "log"

        itsdangerous "github.com/iromli/go-itsdangerous"
)

func main() {
        secret := "seeecret"
        msg := "6858"
        signature := itsdangerous.NewSignature(secret, "", "", "", nil, nil)
        signed_msg, _ := signature.Sign(msg)
        log.Println(signed_msg)
}
```

```
panic: crypto/hmac: hash generation function does not produce unique values

goroutine 1 [running]:
crypto/hmac.New(0xc000068e28, 0xc000014048, 0x14, 0x18, 0x14, 0x18)
	/usr/local/go-faketime/src/crypto/hmac/hmac.go:143 +0x370
github.com/iromli/go-itsdangerous.(*HMACAlgorithm).GetSignature(0xc000010280, 0xc000014030, 0x14, 0x4ce4e6, 0x4, 0x7f346845a108, 0x60, 0xc000056180)
	/tmp/gopath230749032/pkg/mod/github.com/iromli/go-itsdangerous@v0.0.0-20190903235033-828af6836b4f/algorithm.go:23 +0xa5
github.com/iromli/go-itsdangerous.(*Signature).Get(0xc000056180, 0x4ce4e6, 0x4, 0x100000000000001, 0xc000056180, 0xc000068ef0, 0x4a7b0d)
	/tmp/gopath230749032/pkg/mod/github.com/iromli/go-itsdangerous@v0.0.0-20190903235033-828af6836b4f/signature.go:68 +0x99
github.com/iromli/go-itsdangerous.(*Signature).Sign(0xc000056180, 0x4ce4e6, 0x4, 0x13, 0x4ce3a7, 0x1, 0x4cf946)
	/tmp/gopath230749032/pkg/mod/github.com/iromli/go-itsdangerous@v0.0.0-20190903235033-828af6836b4f/signature.go:88 +0x45
main.main()
	/tmp/sandbox285037293/prog.go:13 +0x79

Program exited: status 2.
```

--- a/algorithm.go
+++ b/algorithm.go
@@ -14,13 +14,13 @@ type SigningAlgorithm interface {
 
 // HMACAlgorithm provides signature generation using HMACs.
 type HMACAlgorithm struct {
-	DigestMethod hash.Hash
+	DigestMethod func() hash.Hash
 }
 
 // GetSignature returns the signature for the given key and value.
 func (a *HMACAlgorithm) GetSignature(key, value string) []byte {
-	a.DigestMethod.Reset()
-	h := hmac.New(func() hash.Hash { return a.DigestMethod }, []byte(key))
+	a.DigestMethod().Reset()
+	h := hmac.New(func() hash.Hash { return a.DigestMethod() }, []byte(key))
 	h.Write([]byte(value))
 	return h.Sum(nil)
 }

--- a/go.mod
+++ b/go.mod
@@ -1,3 +1,3 @@
-module github.com/iromli/go-itsdangerous
+module github.com/czekan/go-itsdangerous
 
-go 1.12
+go 1.16

--- a/signature.go
+++ b/signature.go
@@ -24,7 +24,7 @@ type Signature struct {
 	Sep           string
 	Salt          string
 	KeyDerivation string
-	DigestMethod  hash.Hash
+	DigestMethod  func() hash.Hash
 	Algorithm     SigningAlgorithm
 }
 
@@ -35,19 +35,19 @@ func (s *Signature) DeriveKey() (string, error) {
 	var key string
 	var err error
 
-	s.DigestMethod.Reset()
+	s.DigestMethod().Reset()
 
 	switch s.KeyDerivation {
 	case "concat":
-		h := s.DigestMethod
+		h := s.DigestMethod()
 		h.Write([]byte(s.Salt + s.SecretKey))
 		key = string(h.Sum(nil))
 	case "django-concat":
-		h := s.DigestMethod
+		h := s.DigestMethod()
 		h.Write([]byte(s.Salt + "signer" + s.SecretKey))
 		key = string(h.Sum(nil))
 	case "hmac":
-		h := hmac.New(func() hash.Hash { return s.DigestMethod }, []byte(s.SecretKey))
+		h := hmac.New(func() hash.Hash { return s.DigestMethod() }, []byte(s.SecretKey))
 		h.Write([]byte(s.Salt))
 		key = string(h.Sum(nil))
 	case "none":
@@ -108,7 +108,7 @@ func (s *Signature) Unsign(signed string) (string, error) {
 }
 
 // NewSignature creates a new Signature
-func NewSignature(secret, salt, sep, derivation string, digest hash.Hash, algo SigningAlgorithm) *Signature {
+func NewSignature(secret, salt, sep, derivation string, digest func() hash.Hash, algo SigningAlgorithm) *Signature {
 	if salt == "" {
 		salt = "itsdangerous.Signer"
 	}
@@ -119,7 +119,7 @@ func NewSignature(secret, salt, sep, derivation string, digest hash.Hash, algo S
 		derivation = "django-concat"
 	}
 	if digest == nil {
-		digest = sha1.New()
+		digest = sha1.New
 	}
 	if algo == nil {
 		algo = &HMACAlgorithm{DigestMethod: digest}
@@ -194,7 +194,7 @@ func (s *TimestampSignature) Unsign(value string, maxAge uint32) (string, error)
 }
 
 // NewTimestampSignature creates a new TimestampSignature
-func NewTimestampSignature(secret, salt, sep, derivation string, digest hash.Hash, algo SigningAlgorithm) *TimestampSignature {
+func NewTimestampSignature(secret, salt, sep, derivation string, digest func() hash.Hash, algo SigningAlgorithm) *TimestampSignature {
 	s := NewSignature(secret, salt, sep, derivation, digest, algo)
 	return &TimestampSignature{Signature: *s}
 }

